pkg/kv/kvclient/kvcoord/txncorrectnesstest/txncorrectnesstest_test: TestTxnDBReadSkewAnomaly failed

pkg/kv/kvclient/kvcoord/txncorrectnesstest/txncorrectnesstest_test.TestTxnDBReadSkewAnomaly [failed](https://mesolite.cluster.engflow.com/invocations/default/471b48cb-ef33-4d10-8a87-d5b58604e0a6?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZjbGllbnQva3Zjb29yZC90eG5jb3JyZWN0bmVzc3Rlc3Q6dHhuY29ycmVjdG5lc3N0ZXN0X3Rlc3Q=) on master @ [b5debe9e969f3b23e133510abd5c013d52c5a515](https://github.com/cockroachdb/cockroach/commits/b5debe9e969f3b23e133510abd5c013d52c5a515):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x45fcd77]

```
Stack: 

```
goroutine 37960 [running]:
github.com/cockroachdb/cockroach/pkg/kv/kvserver/mmaintegration.(*AllocatorSync).GetMMAStoreStatuses(0x0)
	pkg/kv/kvserver/mmaintegration/allocator_sync.go:286 +0x37
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).rebalance(0xc001e21380, {0x9ebfd08, 0xc001d2adc0}, 0x1)
	pkg/kv/kvserver/mma_store_rebalancer.go:137 +0xd9
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).run(0xc001e21380, {0x9ebfd08, 0xc001d2adc0}, 0xc0040e1b80)
	pkg/kv/kvserver/mma_store_rebalancer.go:96 +0x490
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).start.func1({0x9ebfd08, 0xc001d2adc0})
	pkg/kv/kvserver/mma_store_rebalancer.go:125 +0x4d
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x9ebfd08, 0xc001d2adc0})
	pkg/util/stop/stopper.go:447 +0x113
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 37882
	pkg/util/stop/stopper.go:445 +0x247
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTxnDBReadSkewAnomaly
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestTxnDBReadSkewAnomaly2149842829
    test_log_scope.go:82: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnDBReadSkewAnomaly.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63779

pkg/kv/kvclient/kvcoord/txncorrectnesstest/txncorrectnesstest_test.TestTxnDBReadSkewAnomaly [failed](https://mesolite.cluster.engflow.com/invocations/default/56f502f1-9173-4025-96e5-5516e50a5999?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZjbGllbnQva3Zjb29yZC90eG5jb3JyZWN0bmVzc3Rlc3Q6dHhuY29ycmVjdG5lc3N0ZXN0X3Rlc3Q=) on master @ [33d214cbfa76c1b48c0ce83f55bf7dddc809107b](https://github.com/cockroachdb/cockroach/commits/33d214cbfa76c1b48c0ce83f55bf7dddc809107b):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x45fd597]

```
Stack: 

```
goroutine 37935 [running]:
github.com/cockroachdb/cockroach/pkg/kv/kvserver/mmaintegration.(*AllocatorSync).GetMMAStoreStatuses(0x0)
	pkg/kv/kvserver/mmaintegration/allocator_sync.go:286 +0x37
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).rebalance(0xc00229b5c0, {0x9ec6708, 0xc005823040}, 0x1)
	pkg/kv/kvserver/mma_store_rebalancer.go:137 +0xd9
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).run(0xc00229b5c0, {0x9ec6708, 0xc005823040}, 0xc003c42fa0)
	pkg/kv/kvserver/mma_store_rebalancer.go:96 +0x490
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).start.func1({0x9ec6708, 0xc005823040})
	pkg/kv/kvserver/mma_store_rebalancer.go:125 +0x4d
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x9ec6708, 0xc005823040})
	pkg/util/stop/stopper.go:447 +0x113
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 37857
	pkg/util/stop/stopper.go:445 +0x247
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTxnDBReadSkewAnomaly
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestTxnDBReadSkewAnomaly59337989
    test_log_scope.go:82: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnDBReadSkewAnomaly.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvserver/mma_store_rebalancer.go
+++ b/pkg/kv/kvserver/mma_store_rebalancer.go
@@ -18,6 +18,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/mmaintegration"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -59,11 +60,23 @@ func newMMAStoreRebalancer(
 	// Initialize disk utilization thresholds from cluster settings.
 	opts := allocatorimpl.MakeDiskCapacityOptions(&st.SV)
 	mma.SetDiskUtilThresholds(opts.RebalanceToThreshold, opts.ShedAndBlockAllThreshold)
+	as := s.cfg.AllocatorSync
+	if as == nil {
+		// Production code (server.go) is expected to populate AllocatorSync on
+		// the StoreConfig. A handful of test entry points (TestStoreConfig and
+		// the assorted store_rebalancer_test setups) populate StorePool but not
+		// AllocatorSync; fall back to constructing one here so those tests don't
+		// trip a nil-receiver panic from the background rebalancer goroutine.
+		if !buildutil.CrdbTestBuild {
+			panic(errors.AssertionFailedf("AllocatorSync must be set on StoreConfig outside of test builds"))
+		}
+		as = mmaintegration.NewAllocatorSync(s.cfg.StorePool, mma, st, nil /* knobs */)
+	}
 	return &mmaStoreRebalancer{
 		store:              (*mmaStore)(s),
 		mma:                mma,
 		st:                 st,
-		as:                 s.cfg.AllocatorSync,
+		as:                 as,
 		processTimeoutFunc: makeRateLimitedTimeoutFunc(rebalanceSnapshotRate),
 	}
 }

--- a/pkg/testutils/localtestcluster/BUILD.bazel
+++ b/pkg/testutils/localtestcluster/BUILD.bazel
@@ -22,6 +22,7 @@ go_library(
         "//pkg/kv/kvserver/kvstorage",
         "//pkg/kv/kvserver/liveness",
         "//pkg/kv/kvserver/load",
+        "//pkg/kv/kvserver/mmaintegration",
         "//pkg/kv/kvserver/storeliveness",
         "//pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer",
         "//pkg/roachpb",

--- a/pkg/testutils/localtestcluster/local_test_cluster.go
+++ b/pkg/testutils/localtestcluster/local_test_cluster.go
@@ -27,6 +27,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvstorage"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/load"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/mmaintegration"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/storeliveness"
 	"github.com/cockroachdb/cockroach/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
@@ -249,6 +250,8 @@ func (ltc *LocalTestCluster) Start(t testing.TB, initFactory InitFactoryFn) {
 	)
 	cfg.MMAllocator = mmaprototype.NewAllocatorState(timeutil.DefaultTimeSource{},
 		rand.New(rand.NewSource(timeutil.Now().UnixNano())))
+	cfg.AllocatorSync = mmaintegration.NewAllocatorSync(
+		cfg.StorePool, cfg.MMAllocator, cfg.Settings, nil /* knobs */)
 
 	cfg.Transport = kvserver.NewDummyRaftTransport(cfg.AmbientCtx, cfg.Settings, ltc.Clock)
 	cfg.ClosedTimestampReceiver = sidetransport.NewReceiver(nc, ltc.stopper, ltc.Stores, nil /* testingKnobs */)

